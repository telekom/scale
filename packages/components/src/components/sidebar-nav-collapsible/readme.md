# scale-sidebar-nav-collapsible

We tried adding ::-moz-focus-inner for the button but JSS fails to parse that selector.

- https://developer.mozilla.org/en-US/docs/Web/CSS/::-moz-focus-inner
- https://stackoverflow.com/a/199319

<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                                                 | Type      | Default     |
| -------------- | --------------- | --------------------------------------------------------------------------- | --------- | ----------- |
| `active`       | `active`        | Label and icon get the active color                                         | `boolean` | `false`     |
| `bold`         | `bold`          | Bold label and icon                                                         | `boolean` | `false`     |
| `condensed`    | `condensed`     | Used normally for third level items                                         | `boolean` | `false`     |
| `expanded`     | `expanded`      | Set this to `true` to expand                                                | `boolean` | `undefined` |
| `href`         | `href`          | The URL where the link should point to                                      | `string`  | `'#'`       |
| `label`        | `label`         | The text for the button                                                     | `string`  | `undefined` |
| `nestingLevel` | `nesting-level` | Nesting level within the <scale-sidebar-nav> parent, gets set automatically | `number`  | `undefined` |
| `tag`          | `tag`           | The parent wrapper                                                          | `string`  | `'li'`      |


## Shadow Parts

| Part        | Description |
| ----------- | ----------- |
| `"icon"`    |             |
| `"list"`    |             |
| `"wrapper"` |             |


## Dependencies

### Depends on

- [scale-icon-navigation-collapse-down](../icons/navigation-collapse-down)

### Graph
```mermaid
graph TD;
  scale-sidebar-nav-collapsible --> scale-icon-navigation-collapse-down
  style scale-sidebar-nav-collapsible fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
