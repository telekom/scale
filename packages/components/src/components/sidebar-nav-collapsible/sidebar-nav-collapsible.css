/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --opacity-chevron: 1;
  --border-bottom-color: var(--telekom-color-ui-additional-subtle);
  --border-left-color-third-nesting: var(--telekom-color-ui-additional-strong);
  --color-active: var(--telekom-color-text-and-icon-primary-pressed);
  --color-hover: var(--telekom-color-text-and-icon-primary-hovered);
  --color-primary: var(--telekom-color-text-and-icon-primary-hovered);
  --background-color-before-active: var(--color-primary);
  --font-weight-bold: var(--telekom-typography-font-weight-bold);
  --box-shadow-focus: inset 0 0 0 var(--telekom-line-weight-highlight)
    var(--telekom-color-functional-focus);
}
.sidebar-nav-collapsible {
  margin: 0;
  border-bottom-color: var(--border-bottom-color);
  border-bottom-style: solid;
  border-bottom-width: 0;
}

.sidebar-nav-collapsible--condensed {
  border-bottom-width: 1px;
}

.sidebar-nav-collapsible__wrapper {
  border-bottom-color: var(--border-bottom-color);
  border-bottom-style: solid;
  border-bottom-width: 1px;
}

.sidebar-nav-collapsible--condensed .sidebar-nav-collapsible__wrapper {
  border-bottom-width: 0;
}

.sidebar-nav-collapsible__button {
  color: currentColor;
  width: 100%;
  display: flex;
  box-sizing: border-box;
  text-align: left;
  align-items: center;
  padding-top: 1rem;
  padding-right: calc(0.5 * var(--spacing-indent));
  padding-bottom: 1rem;
  padding-left: var(--spacing-indent);
  justify-content: space-between;
  text-decoration: none;
  font-family: inherit;
  border-radius: var(--telekom-radius-small);
}

.sidebar-nav-collapsible__button:hover {
  color: var(--color-hover);
}

.sidebar-nav-collapsible__button:active {
  color: var(--color-active);
}

.sidebar-nav-collapsible__button:focus {
  outline: none;
  box-shadow: var(--box-shadow-focus);
}

.sidebar-nav-collapsible--active .sidebar-nav-collapsible__button {
  position: relative;
  color: var(--color-primary);
}

.sidebar-nav-collapsible--active .sidebar-nav-collapsible__button::before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: var(--left-current-border, 0);
  width: 0; /* The transparent border makes the 2px */
  height: 100%;
  background-color: var(--color-primary);
  /* Accessibility: extra transparent line for Windows HCM */
  border-left: 2px solid transparent;
}

:host([nesting-level='1']) .sidebar-nav-collapsible__button,
.sidebar-nav-collapsible--bold .sidebar-nav-collapsible__button {
  font-weight: var(--font-weight-bold);
}

:host([nesting-level='2']) .sidebar-nav-collapsible__button {
  padding-left: calc(var(--spacing-indent) + 0.75rem);
}

.sidebar-nav-collapsible__icon {
  transition: opacity 150ms;
}

[aria-expanded='true'] .sidebar-nav-collapsible__icon {
  transform: rotate(0.5turn);
}

@media (hover: hover) {
  [aria-expanded='false'] .sidebar-nav-collapsible__icon {
    opacity: var(--opacity-chevron, 1);
  }

  .sidebar-nav-collapsible__button:hover .sidebar-nav-collapsible__icon {
    opacity: 1;
  }

  .sidebar-nav-collapsible__button:focus .sidebar-nav-collapsible__icon {
    opacity: 1;
  }
}

.sidebar-nav-collapsible__list {
  list-style: none;
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
  border-bottom-width: 1px; /* ? */
}

.sidebar-nav-collapsible--condensed .sidebar-nav-collapsible__list {
  margin-top: 1em;
  margin-left: var(--spacing-indent);
  margin-bottom: 2em;
}
