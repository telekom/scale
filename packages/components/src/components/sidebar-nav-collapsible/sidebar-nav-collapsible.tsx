/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, h, Prop, Host, Element, Watch } from '@stencil/core';
import classNames from 'classnames';

@Component({
  tag: 'scale-sidebar-nav-collapsible',
  styleUrl: 'sidebar-nav-collapsible.css',
  shadow: true,
})
export class SidebarNavCollapsible {
  @Element() el: HTMLElement;

  /** The parent wrapper */
  @Prop() tag?: string = 'li';
  /** The text for the button */
  @Prop() label: string;
  /** The URL where the link should point to */
  @Prop() href: string = '#';
  /** Set this to `true` to expand */
  @Prop({ mutable: true, reflect: true }) expanded: boolean;
  /** Label and icon get the active color */
  @Prop() active?: boolean = false;
  /** Bold label and icon */
  @Prop() bold: boolean = false;
  /** Used normally for third level items */
  @Prop() condensed: boolean = false;
  /** Nesting level within the <scale-sidebar-nav> parent, gets set automatically */
  @Prop() nestingLevel: number;
  /** (optional) Extra styles */
  @Prop() styles?: string;

  @Watch('nestingLevel')
  nestingLevelChanged(newValue: number) {
    if (newValue === 1) {
      this.bold = true;
    }
  }

  handleClick = (event: MouseEvent) => {
    event.preventDefault();
    this.expanded = !this.expanded;
  };

  /**
   * Simulate a <button> allowing using the Space key for toggling the menu.
   */
  handleKeydown = (event: KeyboardEvent) => {
    if (event.metaKey || event.ctrlKey || event.shiftKey) {
      return;
    }
    if (event.defaultPrevented) {
      return;
    }
    if (event.code === 'Space') {
      this.expanded = !this.expanded;
    }
  };

  render() {
    const Tag = this.tag;

    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <Tag
          part={this.getBasePartMap()}
          class={this.getCssClassMap()}
          role="listitem"
        >
          <div class="sidebar-nav-collapsible__wrapper" part="wrapper">
            <a
              href={this.href}
              class="sidebar-nav-collapsible__button"
              onClick={this.handleClick}
              onKeyDown={this.handleKeydown}
              role="button"
              aria-expanded={this.expanded ? 'true' : 'false'}
              part={classNames('button', this.active && 'button-active')}
            >
              {this.label}
              <scale-icon-navigation-collapse-down
                class="sidebar-nav-collapsible__icon"
                selected={this.bold}
                size={20}
                part="icon"
              />
            </a>
          </div>
          <ul
            hidden={!this.expanded}
            class="sidebar-nav-collapsible__list"
            part="list"
          >
            <slot />
          </ul>
        </Tag>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'sidebar-nav-collapsible';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      component,
      this.condensed && `${prefix}condensed`,
      this.active && `${prefix}active`
    );
  }
}
