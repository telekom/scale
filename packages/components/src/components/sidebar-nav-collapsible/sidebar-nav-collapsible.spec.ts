/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { SidebarNavCollapsible } from './sidebar-nav-collapsible';

describe('SidebarNavCollapsible', () => {
  let page: SpecPage;

  describe('snapshots', async () => {
    beforeEach(async () => {
      page = await newSpecPage({
        components: [SidebarNavCollapsible],
        html: `<scale-sidebar-nav-collapsible></scale-sidebar-nav-collapsible>`,
      });
    });
    it('smoke test', async () => {
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('emitter', () => {
    beforeEach(async () => {
      page = await newSpecPage({
        components: [SidebarNavCollapsible],
        html: `<scale-sidebar-nav-collapsible></scale-sidebar-nav-collapsible>`,
      });
    });

    it('should set expanded to false onClick', async () => {
      page.rootInstance.expanded = true;
      const sidebarNavCollapsible = page.root.shadowRoot.querySelector(
        '.sidebar-nav-collapsible__button'
      );
      await page.waitForChanges();
      sidebarNavCollapsible.dispatchEvent(new Event('click'));
      await page.waitForChanges();
      expect(page.rootInstance.expanded).toBe(false);
    });

    it('exit handleKeydown if event metaKey', async () => {
      const sidebarNavCollapsible = page.root.shadowRoot.querySelector(
        '.sidebar-nav-collapsible__button'
      );
      await page.waitForChanges();
      sidebarNavCollapsible.dispatchEvent(
        new KeyboardEvent('keydown', {
          view: window,
          bubbles: true,
          cancelable: true,
          metaKey: true,
        })
      );
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('toggle expanded if event.code = "Space"', async () => {
      page.rootInstance.expanded = true;
      const sidebarNavCollapsible = page.root.shadowRoot.querySelector(
        '.sidebar-nav-collapsible__button'
      );
      await page.waitForChanges();
      sidebarNavCollapsible.dispatchEvent(
        new KeyboardEvent('keydown', {
          view: window,
          bubbles: true,
          cancelable: true,
          code: 'Space',
        })
      );
      await page.waitForChanges();
      expect(page.rootInstance.expanded).toBe(false);
    });
  });

  describe('functions', () => {
    it('nestingLevelChanged turns bold to true', async () => {
      const element = new SidebarNavCollapsible();
      element.bold = false;
      element.nestingLevelChanged(1);
      expect(element.bold).toBe(true);
    });
  });

  describe('has css classes -> getCssClassMap()', () => {
    it('has class component', () => {
      const element = new SidebarNavCollapsible();
      expect(element.getCssClassMap()).toContain('sidebar-nav-collapsible');
    });

    it('has class sidebar-nav-collapsible--condensed', () => {
      const element = new SidebarNavCollapsible();
      element.condensed = true;
      expect(element.getCssClassMap()).toContain(
        'sidebar-nav-collapsible--condensed'
      );
    });

    it('has class sidebar-nav-collapsible--active', () => {
      const element = new SidebarNavCollapsible();
      element.active = true;
      expect(element.getCssClassMap()).toContain(
        'sidebar-nav-collapsible--active'
      );
    });
  });

  describe('has css classes -> getBasePartMap()', () => {
    it('has class component', () => {
      const element = new SidebarNavCollapsible();
      expect(element.getBasePartMap()).toContain('sidebar-nav-collapsible');
    });

    it('has class sidebar-nav-collapsible--condensed', () => {
      const element = new SidebarNavCollapsible();
      element.condensed = true;
      expect(element.getBasePartMap()).toContain('condensed');
    });

    it('has class sidebar-nav-collapsible--active', () => {
      const element = new SidebarNavCollapsible();
      element.active = true;
      expect(element.getBasePartMap()).toContain('active');
    });
  });
});
