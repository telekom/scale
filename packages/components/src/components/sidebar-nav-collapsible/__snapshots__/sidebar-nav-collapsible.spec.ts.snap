// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarNavCollapsible emitter exit handleKeydown if event metaKey 1`] = `
<scale-sidebar-nav-collapsible>
  <mock:shadow-root>
    <li class="sidebar-nav-collapsible" part="sidebar-nav-collapsible" role="listitem">
      <div class="sidebar-nav-collapsible__wrapper" part="wrapper">
        <a aria-expanded="false" class="sidebar-nav-collapsible__button" href="#" part="button" role="button">
          <scale-icon-navigation-collapse-down class="sidebar-nav-collapsible__icon" part="icon" size="16"></scale-icon-navigation-collapse-down>
        </a>
      </div>
      <ul class="sidebar-nav-collapsible__list" hidden="" part="list">
        <slot></slot>
      </ul>
    </li>
  </mock:shadow-root>
</scale-sidebar-nav-collapsible>
`;

exports[`SidebarNavCollapsible snapshots smoke test 1`] = `
<scale-sidebar-nav-collapsible>
  <mock:shadow-root>
    <li class="sidebar-nav-collapsible" part="sidebar-nav-collapsible" role="listitem">
      <div class="sidebar-nav-collapsible__wrapper" part="wrapper">
        <a aria-expanded="false" class="sidebar-nav-collapsible__button" href="#" part="button" role="button">
          <scale-icon-navigation-collapse-down class="sidebar-nav-collapsible__icon" part="icon" size="16"></scale-icon-navigation-collapse-down>
        </a>
      </div>
      <ul class="sidebar-nav-collapsible__list" hidden="" part="list">
        <slot></slot>
      </ul>
    </li>
  </mock:shadow-root>
</scale-sidebar-nav-collapsible>
`;
